pennylane/templates/subroutines/uccsd_unitary.py

```diff
@@ -0,0 +1,231 @@
+                return qml.expval(qml.PauliZ(0))
+
+            params = np.random.normal(0, np.pi, len(ph) + len(pphh))
+            print(circuit(params, wires, ph=ph, pphh=pphh, init_state=ref_state))
+
+    """
+
+    ##############
+    # Input checks
+
+    check_no_variable(wires, msg="'wires' cannot be differentiable")
+
+    wires = check_wires(wires)
+
+    if (not ph) and (not pphh):
+        raise ValueError(
+            "'ph' and 'pphh' lists can not be both empty; got ph={}, pphh={}".format(ph, pphh)
+        )
+
+    check_type(ph, [list], msg="'ph' must be a list; got {}".format(ph))
+    for i_ph in ph:
+        check_type(i_ph, [list], msg="Each element of 'ph' must be a list; got {}".format(i_ph))
+        for i in i_ph:
+            check_type(
+                i, [int], msg="Each element of 'ph' must be a list of integers; got {}".format(i_ph)
+            )
+
+    check_type(pphh, [list], msg="'pphh' must be a list; got {}".format(pphh))
+    for i_pphh in pphh:
+        check_type(
+            i_pphh, [list], msg="Each element of 'pphh' must be a list; got {}".format(i_pphh)
+        )
+        for i in i_pphh:
+            check_type(
+                i,
+                [int],
+                msg="Each element of 'pphh' must be a list of integers; got {}".format(i_pphh),
+            )
+
+    check_type(
+        init_state,
+        [np.ndarray],
+        msg="'init_state' must be a Numpy array; got {}".format(init_state),
+    )
+    for i in init_state:
+        check_type(
+            i,
+            [int, np.int64],
+            msg="Elements of 'init_state' must be integers; got {}".format(init_state),
+        )
+
+    expected_shape = (2,)
+    for i_ph in ph:
+        check_shape(
+            i_ph,
+            expected_shape,
+            msg="Elements of 'ph' must be of shape {}; got {}".format(
+                expected_shape, get_shape(i_ph)
+            ),
+        )
+
+    expected_shape = (4,)
+    for i_pphh in pphh:
+        check_shape(
+            i_pphh,
+            expected_shape,
+            msg="Elements of 'pphh' must be of shape {}; got {}".format(
+                expected_shape, get_shape(i_pphh)
+            ),
+        )
+
+    expected_shape = (len(ph) + len(pphh),)
+    check_shape(
+        weights,
+        expected_shape,
+        msg="'weights' must be of shape {}; got {}".format(expected_shape, get_shape(weights)),
+    )
+
+    expected_shape = (len(wires),)
+    check_shape(
+        init_state,
+        expected_shape,
+        msg="'init_state' must be of shape {}; got {}".format(
+            expected_shape, get_shape(init_state)
+        ),
+    )
+
+    ###############
+
+    qml.BasisState(np.flip(init_state), wires=wires)
+
+    for d, i_pphh in enumerate(pphh):
+        DoubleExcitationUnitary(weights[len(ph) + d], wires=i_pphh)
+
+    for s, i_ph in enumerate(ph):
+        SingleExcitationUnitary(weights[s], wires=i_ph)
```

Comment:
The actual implementation is so much shorter than the input checks 😆

Reply:
Right!.